8Knot/pages/codebase/visualizations/reviewer_file_heatmap.py

```diff
@@ -0,0 +1,455 @@
+    directories = sorted(directories)
+
+    # add top level directory to the list of directories
+    directories.insert(0, "Top Level Directory")
+    logging.warning(f"DIRECTORY DROPDOWN - FINISHED")
+
+    return directories, "Top Level Directory"
+
+
+# callback for reviewer file heatmap graph
+@callback(
+    Output(f"{PAGE}-{VIZ_ID}", "figure"),
+    [
+        Input(f"repo-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"directory-{PAGE}-{VIZ_ID}", "value"),
+        Input("bot-switch", "value"),
+    ],
+    background=True,
+)
+def reviewer_file_heatmap_graph(repo_id, directory, bot_switch):
+    start = time.perf_counter()
+    logging.warning(f"{VIZ_ID}- START")
+
+    # get dataframes of data from cache
+    df_file, df_actions, df_file_cntbs = multi_query_helper([repo_id])
+
+    # test if there is data
+    if df_file.empty or df_actions.empty or df_file_cntbs.empty:
+        logging.warning(f"{VIZ_ID} - NO DATA AVAILABLE")
+        return nodata_graph
+
+    # function for all data pre processing
+    df = process_data(df_file, df_actions, df_file_cntbs, directory, bot_switch)
+
+    # if there are no cntrbs in a directory plot no data graph
+    if df.empty:
+        return nodata_graph
+
+    fig = create_figure(df)
+
+    logging.warning(f"{VIZ_ID} - END - {time.perf_counter() - start}")
+    return fig
+
+
+def multi_query_helper(repos):
+    """
+    For reviewer_file_heatmap_graph-
+    hack to put all of the cache-retrieval
+    in the same place temporarily
+    """
+
+    # wait for data to asynchronously download and become available.
+    while not_cached := cf.get_uncached(func_name=rfq.__name__, repolist=repos):
+        logging.warning(f"CONTRIBUTOR FILE HEATMAP - WAITING ON DATA TO BECOME AVAILABLE")
+        time.sleep(0.5)
+
+    # wait for data to asynchronously download and become available.
+    while not_cached := cf.get_uncached(func_name=cnq.__name__, repolist=repos):
+        logging.warning(f"CONTRIBUTOR FILE HEATMAP - WAITING ON DATA TO BECOME AVAILABLE")
+        time.sleep(0.5)
+
+    # wait for data to asynchronously download and become available.
+    while not_cached := cf.get_uncached(func_name=cpfq.__name__, repolist=repos):
+        logging.warning(f"CONTRIBUTOR FILE HEATMAP - WAITING ON DATA TO BECOME AVAILABLE")
+        time.sleep(0.5)
+
+    # GET ALL DATA FROM POSTGRES CACHE
+    df_file = cf.retrieve_from_cache(
+        tablename=rfq.__name__,
+        repolist=repos,
+    )
+    df_actions = cf.retrieve_from_cache(
+        tablename=cnq.__name__,
+        repolist=repos,
+    )
+    df_file_cntrbs = cf.retrieve_from_cache(
+        tablename=cpfq.__name__,
+        repolist=repos,
+    )
+
+    # necessary preprocessing steps that were lifted out of the querying step
+    df_actions = preproc_u.contributors_df_action_naming(df_actions)
+    df_file_cntrbs = preproc_u.cntrb_per_file(df_file_cntrbs)
+
+    return df_file, df_actions, df_file_cntrbs
+
+
+def process_data(
```

Comment:
I think this would be easier to review, understand, and maintain if it was broken up into helper functions. I think we should do this to the other two visualizations of this kind retroactively.

Major pillars of the pipeline could be functions that kinda chain together with docstrings that describe what they do.

I find that reading Pandas code after not working on it for a while takes a LOT of mental tracing, which I'd like to minimize.

Reply:
@JamesKunstle I just updated the pr with some potential helper functions. Let me know if that breakdown makes sense/ you have suggestions on other options of chunking the code into helper functions. It was not intuitive to me on how to break it apart. Once we agree with the breakdown ill add docstrings and we can discuss naming more. My goal is to come up with a structure here that can be applied to the others as well on a different pr

Reply:
@cdolfi that'd be fantastic